### PR TITLE
Fix 99 (and allow qualified static linking)

### PR DESCRIPTION
**Pull request by dsyme**
This fixes our oldest "Urgency-Soon" issue here: https://github.com/Microsoft/visualfsharp/issues/99

It also includes some work I did a while back so to extend the static linking support to allow static linking of compoennts ike System.Collections.Immutable.  The addition is that the staticlink option now allows 

    --staticlink:AssembyName,Prefix

where ``Prefix`` is used to disambiguate between various types in statically components that have the same name.  This is becoming a particular problem since all Microsoft components tend to have a private "version" type with the same name. We don't use this in the code base but I'm including it since I also happened to fix  https://github.com/Microsoft/visualfsharp/issues/99 while looking at this problem, and it seems we may as well include the extended support as well.

### CODE REVIEW

File: src/fsharp/CompileOptions.fs [hunk 1]
```diff
@@ -776,7 +776,13 @@ let advancedFlagsFsc tcConfigB =
                                              tcConfigB.implicitlyResolveAssemblies <- true), None,
                              Some (FSComp.SR.optsStandalone()))
 
-        yield CompilerOption("staticlink", tagFile, OptionString (fun s -> tcConfigB.extraStaticLinkRoots <- tcConfigB.extraStaticLinkRoots @ [s]), None,
+        yield CompilerOption("staticlink", tagFile, OptionString (fun s -> 
+                                             if s.StartsWith("-") then 
+                                                 let assName = s.[1..]
```
Comment (KevinRansom) [inline src/fsharp/CompileOptions.fs]: assName?

**Comment by dsyme** [comment]
Closing as I'll resubmit without the new feature (the tests have not been updated)

File: src/fsharp/CompileOptions.fs (hunk 1 above)
Comment (KevinRansom) [inline src/fsharp/CompileOptions.fs]: s.[1..] this generates a ton of inline code .... perhaps we can add an optimization to the compiler to improve code generation?

Amusingly it even calls String.Substring at the end.

this:
````
[<EntryPoint>]
let main argv = 
    let s = "Hello, world"
    let _b = s.[1..]
    0 // return an integer exit code
````
becomes:
````
  .custom instance void [FSharp.Core]Microsoft.FSharp.Core.CompilationMappingAttribute::.ctor(valuetype [FSharp.Core]Microsoft.FSharp.Core.SourceConstructFlags) = ( 01 00 07 00 00 00 00 00 ) 
  .method public static int32  main(string[] argv) cil managed
  {
    .entrypoint
    .custom instance void [FSharp.Core]Microsoft.FSharp.Core.EntryPointAttribute::.ctor() = ( 01 00 00 00 ) 
    // Code size       264 (0x108)
    .maxstack  5
    .locals init ([0] string s,
             [1] string _b,
             [2] string V_2,
             [3] class [FSharp.Core]Microsoft.FSharp.Core.FSharpOption`1<int32> V_3,
             [4] class [FSharp.Core]Microsoft.FSharp.Core.FSharpOption`1<int32> V_4,
             [5] class [mscorlib]System.Tuple`2<int32,int32> V_5,
             [6] int32 V_6,
             [7] class [FSharp.Core]Microsoft.FSharp.Core.FSharpOption`1<int32> V_7,
             [8] int32 V_8,
             [9] class [FSharp.Core]Microsoft.FSharp.Core.FSharpOption`1<int32> V_9,
             [10] int32 V_10,
             [11] class [FSharp.Core]Microsoft.FSharp.Core.FSharpOption`1<int32> V_11,
             [12] class [FSharp.Core]Microsoft.FSharp.Core.FSharpOption`1<int32> V_12,
             [13] int32 V_13,
             [14] int32 V_14,
             [15] int32 V_15,
             [16] int32 V_16,
             [17] int32 V_17)
    IL_0000:  nop
    IL_0001:  ldstr      "Hello, world"
    IL_0006:  stloc.0
    IL_0007:  ldloc.0
    IL_0008:  stloc.2
    IL_0009:  ldc.i4.1
    IL_000a:  call       class [FSharp.Core]Microsoft.FSharp.Core.FSharpOption`1<!0> class [FSharp.Core]Microsoft.FSharp.Core.FSharpOption`1<int32>::Some(!0)
    IL_000f:  stloc.3
    IL_0010:  ldnull
    IL_0011:  stloc.s    V_4
    IL_0013:  ldloc.2
    IL_0014:  callvirt   instance int32 [mscorlib]System.String::get_Length()
    IL_0019:  stloc.s    V_6
    IL_001b:  ldloc.3
    IL_001c:  brtrue.s   IL_005d

    IL_001e:  ldloc.s    V_4
    IL_0020:  brfalse.s  IL_0034

    IL_0022:  ldloc.s    V_4
    IL_0024:  stloc.s    V_7
    IL_0026:  ldloc.s    V_7
    IL_0028:  call       instance !0 class [FSharp.Core]Microsoft.FSharp.Core.FSharpOption`1<int32>::get_Value()
    IL_002d:  ldc.i4.0
    IL_002e:  blt.s      IL_0032

    IL_0030:  br.s       IL_0046

    IL_0032:  br.s       IL_005d

    IL_0034:  ldc.i4.0
    IL_0035:  ldc.i4.0
    IL_0036:  ldloc.s    V_6
    IL_0038:  add
    IL_0039:  ldc.i4.1
    IL_003a:  sub
    IL_003b:  newobj     instance void class [mscorlib]System.Tuple`2<int32,int32>::.ctor(!0,
                                                                                          !1)
    IL_0040:  nop
    IL_0041:  br         IL_00cc

    IL_0046:  ldloc.s    V_7
    IL_0048:  call       instance !0 class [FSharp.Core]Microsoft.FSharp.Core.FSharpOption`1<int32>::get_Value()
    IL_004d:  stloc.s    V_8
    IL_004f:  ldc.i4.0
    IL_0050:  ldloc.s    V_8
    IL_0052:  newobj     instance void class [mscorlib]System.Tuple`2<int32,int32>::.ctor(!0,
                                                                                          !1)
    IL_0057:  nop
    IL_0058:  br         IL_00cc

    IL_005d:  ldloc.3
    IL_005e:  brfalse.s  IL_0091

    IL_0060:  ldloc.3
    IL_0061:  stloc.s    V_9
    IL_0063:  ldloc.s    V_4
    IL_0065:  brtrue.s   IL_0091

    IL_0067:  ldloc.s    V_9
    IL_0069:  call       instance !0 class [FSharp.Core]Microsoft.FSharp.Core.FSharpOption`1<int32>::get_Value()
    IL_006e:  ldc.i4.0
    IL_006f:  ldloc.s    V_6
    IL_0071:  add
    IL_0072:  bgt.s      IL_0076

    IL_0074:  br.s       IL_0078

    IL_0076:  br.s       IL_0091

    IL_0078:  ldloc.s    V_9
    IL_007a:  call       instance !0 class [FSharp.Core]Microsoft.FSharp.Core.FSharpOption`1<int32>::get_Value()
    IL_007f:  stloc.s    V_10
    IL_0081:  ldloc.s    V_10
    IL_0083:  ldc.i4.0
    IL_0084:  ldloc.s    V_6
    IL_0086:  add
    IL_0087:  ldc.i4.1
    IL_0088:  sub
    IL_0089:  newobj     instance void class [mscorlib]System.Tuple`2<int32,int32>::.ctor(!0,
                                                                                          !1)
    IL_008e:  nop
    IL_008f:  br.s       IL_00cc

    IL_0091:  ldloc.3
    IL_0092:  brfalse.s  IL_00c1

    IL_0094:  ldloc.3
    IL_0095:  stloc.s    V_11
    IL_0097:  ldloc.s    V_4
    IL_0099:  brfalse.s  IL_00a1

    IL_009b:  ldloc.s    V_4
    IL_009d:  stloc.s    V_12
    IL_009f:  br.s       IL_00a3

    IL_00a1:  br.s       IL_00c1

    IL_00a3:  ldloc.s    V_12
    IL_00a5:  call       instance !0 class [FSharp.Core]Microsoft.FSharp.Core.FSharpOption`1<int32>::get_Value()
    IL_00aa:  stloc.s    V_13
    IL_00ac:  ldloc.s    V_11
    IL_00ae:  call       instance !0 class [FSharp.Core]Microsoft.FSharp.Core.FSharpOption`1<int32>::get_Value()
    IL_00b3:  stloc.s    V_14
    IL_00b5:  ldloc.s    V_14
    IL_00b7:  ldloc.s    V_13
    IL_00b9:  newobj     instance void class [mscorlib]System.Tuple`2<int32,int32>::.ctor(!0,
                                                                                          !1)
    IL_00be:  nop
    IL_00bf:  br.s       IL_00cc

    IL_00c1:  newobj     instance void [mscorlib]System.IndexOutOfRangeException::.ctor()
    IL_00c6:  call       !!0 [FSharp.Core]Microsoft.FSharp.Core.Operators::Raise<class [mscorlib]System.Tuple`2<int32,int32>>(class [mscorlib]System.Exception)
    IL_00cb:  nop
    IL_00cc:  stloc.s    V_5
    IL_00ce:  ldloc.s    V_5
    IL_00d0:  call       instance !0 class [mscorlib]System.Tuple`2<int32,int32>::get_Item1()
    IL_00d5:  stloc.s    V_15
    IL_00d7:  ldloc.s    V_5
    IL_00d9:  call       instance !1 class [mscorlib]System.Tuple`2<int32,int32>::get_Item2()
    IL_00de:  stloc.s    V_16
    IL_00e0:  ldloc.s    V_16
    IL_00e2:  ldloc.s    V_15
    IL_00e4:  sub
    IL_00e5:  ldc.i4.1
    IL_00e6:  add
    IL_00e7:  stloc.s    V_17
    IL_00e9:  ldloc.s    V_17
    IL_00eb:  ldc.i4.0
    IL_00ec:  bgt.s      IL_00f0

    IL_00ee:  br.s       IL_00f2

    IL_00f0:  br.s       IL_00fa

    IL_00f2:  ldsfld     string [mscorlib]System.String::Empty
    IL_00f7:  nop
    IL_00f8:  br.s       IL_0105

    IL_00fa:  ldloc.2
    IL_00fb:  ldloc.s    V_15
    IL_00fd:  ldloc.s    V_17
    IL_00ff:  callvirt   instance string [mscorlib]System.String::Substring(int32,
                                                                            int32)
    IL_0104:  nop
    IL_0105:  stloc.1
    IL_0106:  ldc.i4.0
    IL_0107:  ret
  } // end of method Program::main

} // end of class Program

````

**Comment by dsyme** [comment]
> Perhaps an additional to the help text noting prefix?

Yeah, one of the reasons I closed was because I lost all the commits that had that sort of thing implemented.  I don't know what happened to them, will try to find them.

**Pull request closed without merging**